internals-js/src/operations.ts

```diff
@@ -1535,6 +1551,32 @@ export class SelectionSet {
     this._selections = mapValues(keyedSelections);
   }
 
+  minimizeSelectionSet(namedFragments: NamedFragments, fragmentSpreadsById: Map<string, FragmentSpreadSelection>): SelectionSet {
+    return this.lazyMap((selection) => {
+      if (selection.kind === 'FragmentSelection' && selection.element.typeCondition && selection.element.appliedDirectives.length === 0 && selection.selectionSet) {
+        const id = print(selection.toSelectionNode());
+        const existing = fragmentSpreadsById.get(id);
+        if (existing) return existing;
+
+        const minimizedSelectionSet = selection.selectionSet.minimizeSelectionSet(namedFragments, fragmentSpreadsById);
+        const fragmentDefinition = new NamedFragmentDefinition(
+          this.parentType.schema(),
+          `qp__${fragmentSpreadsById.size}`,
```

Comment:
I think we could generate some better names -> maybe something along the lines `selection.element.typeCondition.name + "Fragment" + <some unique identifier>`